Classes/ElasticSearchClient.php

```diff
@@ -37,13 +38,94 @@ class ElasticSearchClient extends Client
      */
     protected $indexNameStrategy;
 
+    /**
+     * @var DimensionsService
+     * @Flow\Inject
+     */
+    protected $dimensionsService;
+
+    /**
+     * @var string
+     */
+    protected $dimensionsHash;
+
+    /**
+     * @var NodeInterface
+     */
+    protected $contextNode;
+
+    /**
+     * @return NodeInterface
+     */
+    public function getContextNode(): NodeInterface
+    {
+        return $this->contextNode;
+    }
+
+    /**
+     * @param NodeInterface $contextNode
+     */
+    public function setContextNode(NodeInterface $contextNode): void
+    {
+        $this->setDimensions($contextNode->getContext()->getTargetDimensions());
+        $this->contextNode = $contextNode;
+    }
+
+    /**
+     * @param array $dimensionValues
+     */
+    public function setDimensions(array $dimensionValues = []): void
+    {
+        $this->dimensionsHash = $this->dimensionsService->hash($dimensionValues);
+    }
+
+    /**
+     * @return string
+     */
+    public function getDimensionsHash(): string
+    {
+        return $this->dimensionsHash;
+    }
+
+    /**
+     * @param \Closure $closure
+     * @param array $dimensionValues
+     * @throws \Exception
+     */
+    public function withDimensions(\Closure $closure, array $dimensionValues = []): void
+    {
+        $previousDimensionHash = $this->dimensionsHash;
+        try {
+            $this->setDimensions($dimensionValues);
+            $closure();
+            $this->dimensionsHash = $previousDimensionHash;
+        } catch (\Exception $exception) {
```

Comment:
You could use a finally block that contains the dimensionsHash-Reset; that avoids code duplication and you do not need to re-throw a catched exception.